### PR TITLE
Regularizers

### PR DESCRIPTION
**Pull request by hunse**
Add `Percentile` and `PercentileRange` regularizers.

### CODE REVIEW

**Comment by drasmuss** [comment]
Pushed a fixup with a couple changes:

- Keep regularizers in the `keras_spiking.regularizers` name space. This is designed to align with the Keras API, where all the regularizers are organized under the `keras.regularizers` name-space. One could argue that then the layers should be under `keras_spiking.layers`, but since those are the main user-facing access points of this package that seemed too awkward, so convenience outweighted consistency. But since the regularizers aren't as central a part of the package, I'd favour consistency with the Keras organization in this case.
- Separate the notion of applying regularization to a range from the Percentile class into its own class, and then share that functionality across all other regularizers. This also removes the need for separate Percentile/PercentileRange classes. This also simplifies the Percentile API/implementation somewhat (since we're offloading the "secondary" regularization, e.g. L1/L2, to an already implemented class, rather than re-implementing that within Percentile).
- Simplified the Percentile example a bit, by directly creating a functional model. I also found that setting `spiking_aware_training=False` didn't seem to improve the results, so I just left it at the default (True) to make things simpler.

File: keras_spiking/regularizers.py [hunk 1]
```diff
@@ -16,27 +80,34 @@ class L1L2(tf.keras.regularizers.L1L2):
...
-        this point).
+    target : float or tuple
+        The value that we want the regularized outputs to be driven towards. Can be a
+        float, in which case all outputs will be driven towards that value, or a tuple
+        specifying a range ``(min, max)``, in which case outputs outside that range
+        will be driven towards that range (but outputs within the range will not be
+        penalized).
     """
 
     def __init__(self, l1=0.0, l2=0.0, target=0, **kwargs):
```
Comment (hunse) [inline keras_spiking/regularizers.py:91]: Should we have a warning if both `l1` and `l2` are 0? I could see users not changing these (since there are defaults), and being surprised when nothing happens.

File: keras_spiking/regularizers.py [hunk 2]
```diff
@@ -65,137 +139,127 @@ class L2(L1L2):
...
+       MAE: ...
     """
 
     def __init__(
-        self, scale=0.01, minimum=0, maximum=0, percentile=100, axis=0, penalty="l2"
+        self,
+        percentile=100,
+        axis=0,
+        target=0,
+        regularizer=tf.keras.regularizers.L2(0.01),
```
Comment (hunse) [inline keras_spiking/regularizers.py:244]: This does make it more annoying to change the weight on the regularization, since the user has to pass a full regularizer class. I'm wondering if we should just force this to be `L1L2` and then make `l1` and `l2` as parameters to `Percentile`.

File: keras_spiking/regularizers.py (hunk 1 above)
Comment (drasmuss) [reply keras_spiking/regularizers.py:91]: This is designed to mimic https://www.tensorflow.org/api_docs/python/tf/keras/regularizers/L1L2, where they both default to 0

Comment (drasmuss) [reply keras_spiking/regularizers.py:91]: Oh nm you were just saying add a warning, not change the defaults. Yeah that sounds good :+1:

File: keras_spiking/regularizers.py (hunk 2 above)
Comment (drasmuss) [reply keras_spiking/regularizers.py:244]: Yeah that sounds good. I think it's pretty uncommon for people to use something other than L1/L2, so the added convenience is probably worth the loss of that small potential functionality.

File: keras_spiking/regularizers.py [hunk 3]
```diff
@@ -65,137 +139,127 @@ class L2(L1L2):
...
     .. testoutput::
        :hide:
-       :options: +ELLIPSIS, +NORMALIZE_WHITESPACE
 
        ...
        Max rates: ...
-       Error: ...
-
-    We use ``spiking_aware_training=False``, since activity regularizers perform better
-    on continuous activations (firing rates) than they do on discrete spikes.
```
Comment (hunse) [inline keras_spiking/regularizers.py]: In the case of this example, you found that it worked fine with `spiking_aware_training=True` works fine, but in general it doesn't. I ran into a number of cases where the optimization wasn't converging with it on. Also, the loss will be much higher with it on, because even if the target rate is 500 Hz and the neuron is spiking every second timestep, the output values will be [1000, 0, 1000, 0] which will be penalized.

In general, we really do want to be averaging over time when we apply regularizers, I think, if `spiking_aware_training=True`. Without something to average over time, I think it's best to recommend keeping it off with regularizers. I've found regularizers to be really finicky for changing spike rates (partly to do with the lack of special initialization to get firing rates in the right ballpark), and I'm worried that users would take this example and put it into their own code and find it doesn't work. Even something as simple as changing the amount of regularization slightly can cause it to not converge.

File: keras_spiking/regularizers.py [hunk 4]
```diff
@@ -65,137 +139,127 @@ class L2(L1L2):
...
+            dt=1,
+            activity_regularizer=keras_spiking.regularizers.Percentile(target=(5, 10)),
+        )(x)
+        x = tf.keras.layers.GlobalAveragePooling1D()(x)
+        x = tf.keras.layers.Dense(1)(x)
+
+        model = tf.keras.Model(inp, (x, spikes))
+
+        model.compile(
+            optimizer="rmsprop", loss={"dense_1": "mse"}, metrics={"dense_1": "mae"}
```
Comment (hunse) [inline keras_spiking/regularizers.py:216]: Why do we need to specify "dense_1" here? Is that because the model outputs "spikes" too, and without specifying that, it would also apply "mse" to the spikes? If that's the case, I should definitely re-examine my development of this example, because that could have been causing a lot of my problems.

Also, we should really highlight that for users (with a comment at least), since that seems like an easy/common pitfall.

File: keras_spiking/regularizers.py (hunk 3 above)
Comment (drasmuss) [reply keras_spiking/regularizers.py]: I also tried setting `spiking_aware_training=False` in the notebook where we apply L2 regularization, and it made it worse. So I don't think it's clear which option is better in general, which is why I opted to just not bring it up here.

Comment (hunse) [reply keras_spiking/regularizers.py]: In that case, I would still mention the issue here, and say it's unclear which is better and an area of active research (or something like that). I want to make it clear to users that this is not a copy-paste one-size-fits-all example, and for their own networks, they might have to play around.

Maybe this isn't the best place for that, though, and we should add something to the Sphinx docs. We could then point to that here without going into detail.

It would also be good to mention the initialization issue somewhere (i.e. that if you can initialize your network with firing rates that are closer to the target values, optimization can be easier). It *will* make users say "how do you do that?", but I think it's better to have it out in the open, rather than not mention it and then have users frustrated if they try to use our regularizers and they don't work.

Comment (drasmuss) [reply keras_spiking/regularizers.py]: I added a note on both these points to the regularization section in the example (it seemed like too much text to try to fit into the docstring example).

**Comment by drasmuss** [comment]
I think all comments are addressed in fixups, will merge if it looks good to you

**Comment by hunse** [comment]
LGTM!

**Review by drasmuss — APPROVED** [review]
